### PR TITLE
Freq sort exercise

### PR DESCRIPTION
**Pull request by chanchalupadhyay**

### CODE REVIEW

File: js-exercises/freq-sort/freqSort.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
+function freqSort(arrayOfCharacters) {
+  if (!Array.isArray(arrayOfCharacters)) {
+    throw new Error("argument should be array");
+  } else {
```
Comment (venky111) [inline js-exercises/freq-sort/freqSort.js:4]: else not required

File: js-exercises/freq-sort/freqSort.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
+function freqSort(arrayOfCharacters) {
+  if (!Array.isArray(arrayOfCharacters)) {
+    throw new Error("argument should be array");
```
Comment (venky111) [inline js-exercises/freq-sort/freqSort.js:3]: Better to throw TypeError

File: js-exercises/freq-sort/freqSort.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  } else {
+    let frequencyMap = {};
+
+    arrayOfCharacters.map(checkFreqElement);
+
+    function checkFreqElement(element) {
```
Comment (venky111) [inline js-exercises/freq-sort/freqSort.js:9]: Move function outside so it can be reused if required elsewhere

File: js-exercises/freq-sort/freqSort.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+      element = element.normalize("NFC");
+      if (frequencyMap.hasOwnProperty(element))
+        return (frequencyMap[element] += 1);
+      else return (frequencyMap[element] = 1);
```
Comment (venky111) [inline js-exercises/freq-sort/freqSort.js:13]: else is not required

File: js-exercises/freq-sort/freqSort.js (hunk 3 above)
Comment (venky111) [reply js-exercises/freq-sort/freqSort.js:9]: Also here I think function name can be improved as we are not checking frequency we are calculating it. Ex- getfrequencyMapOfArray not exactly but kind of

File: js-exercises/freq-sort/freqSort.js [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    throw new Error("argument should be array");
+  } else {
+    let frequencyMap = {};
+
+    arrayOfCharacters.map(checkFreqElement);
+
+    function checkFreqElement(element) {
+      element = element.normalize("NFC");
+      if (frequencyMap.hasOwnProperty(element))
+        return (frequencyMap[element] += 1);
```
Comment (venky111) [inline js-exercises/freq-sort/freqSort.js:12]: Here instead of using frequencyMap.hasOwnProperty try to use Object.prototype.hasOwnProperty bcz if frequency has a key as 'hasOwnProperty' things will not work as expected, its a very rare thing but its good to do like this.

File: js-exercises/freq-sort/freqSort.js [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+      else return (frequencyMap[element] = 1);
+    }
+
+    let sortedArray = Object.entries(frequencyMap).sort(
+      (firstElement, secondElement) => secondElement[1] - firstElement[1]
+    );
+
```
Comment (venky111) [inline js-exercises/freq-sort/freqSort.js:19]: I think as explained in yesterday's class we can make things declarative here(data transformation style) , making business logic exposed and lang construct hidden

File: js-exercises/freq-sort/freqSort.test.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    expect(freqSort(["a", "z", "z", "z", "b", "b", "z"])).toEqual([
+      "z",
+      "b",
+      "a",
+    ]);
+    expect(
+      freqSort(["b", "z", "z", "z", "b", "b", "z", "c", "h", "h"])
+    ).toEqual(["z", "b", "h", "c"]);
+  });
+  expect(
```
Comment (venky111) [inline js-exercises/freq-sort/freqSort.test.js:14]: Try to cover few edge cases like array having NaN, Undefined,null elements.

**Review by venky111 — COMMENTED** [review]
Hi please find my comments added.
